### PR TITLE
[REVIEW] Initial cuspatial java layer bindings [skip ci]

### PR DESCRIPTION
**Pull request by byin001**

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by harrism** [comment]
OK to test

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # cuSpatial 0.16.0 (Date TBD)
 
 ## New Features
+- PR #290 Add Java binding hence a cuSpatial JAR package for Java users.
```
Comment (harrism) [inline CHANGELOG.md:4]:
```suggestion
- PR #290 Add Java bindings and a cuSpatial JAR package for Java users.
```

**Review by revans2 — CHANGES_REQUESTED** [review]
Overall it looks good.  Mostly I just saw a few nits.

Please note that I am just looking generally at the java and JNI code.  It is almost all boilerplate so I feel fairly confident in this review, but I would not be able to detect some types issues without more of an understanding of the cuspacial APIs themselves.

File: java/pom.xml [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+  limitations under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>ai.rapids</groupId>
+    <artifactId>cuspatial</artifactId>
+    <version>0.15-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (byin001) [reply java/pom.xml:24]: updated to 0.16-SNAPSHOT

File: java/pom.xml [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+                            <goal>execute</goal>
+                        </goals>
+                        <configuration>
+                            <source>
+                            def sout = new StringBuffer(), serr = new StringBuffer()
+                            //This only works on linux
+                            def proc = 'ldd ${native.build.path}/libcuspatialjni.so'.execute()
+                            proc.consumeProcessOutput(sout, serr)
+                            proc.waitForOrKill(10000)
+                            //def librp = ~/librmm\\.so\\s+=>\\s+(.*)librmm.*\\.so\\s+.*/
```
[the remark this answers is not in the document]
Comment (byin001) [reply java/pom.xml:351]: removed

File: java/src/main/java/ai/rapids/cuspatial/CuSpatial.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    long[] columnAddressArray = trajectoryDistancesAndSpeedsImpl(numTrajectories, objectId.getNativeView(), x.getNativeView(), y.getNativeView(), timestamp.getNativeView());
+    return new Table(columnAddressArray);
+  }
+
+  public static Table trajectoryBoundingBoxes(int numTrajectories, ColumnVector objectId, ColumnVector x, ColumnVector y) {
+    long[] columnAddressArray = trajectoryBoundingBoxesImpl(numTrajectories, objectId.getNativeView(), x.getNativeView(), y.getNativeView());
+    return new Table(columnAddressArray);
+  }
+
+  public static ColumnVector pointInPolygon(ColumnVector testPointsX, ColumnVector testPointsY,
```
[the remark this answers is not in the document]
Comment (byin001) [reply java/src/main/java/ai/rapids/cuspatial/CuSpatial.java:118]: javadoc completed

File: java/src/main/java/ai/rapids/cuspatial/CuSpatialException.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
```
[the remark this answers is not in the document]
Comment (byin001) [reply java/src/main/java/ai/rapids/cuspatial/CuSpatialException.java:2]: typo corrected to 2020

File: java/src/main/native/src/cuSpatialJni.cpp [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  for (int i = 0; i < table_cols; i++) {
+    outcol_handles[i] = reinterpret_cast<jlong>(ret[i].release());
+  }
+  for (int i = 0; i < extra_columns.size(); i++) {
+    outcol_handles[i + table_cols] = reinterpret_cast<jlong>(extra_columns[i].release());
+  }
+  return outcol_handles.get_jArray();
+}
+
+jlongArray convert_table_for_return(JNIEnv *env, std::unique_ptr<cudf::table> &table_result) {
```
[the remark this answers is not in the document]
Comment (byin001) [reply java/src/main/native/src/cuSpatialJni.cpp:53]: anonymous namespace added

File: java/src/main/native/src/cuSpatialJni.cpp [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+  std::vector<std::unique_ptr<cudf::column>> extra;
+  return convert_table_for_return(env, table_result, extra);
+}
+
+jlongArray convert_columns_for_return(JNIEnv *env, std::vector<std::unique_ptr<cudf::column>> &columns) {
```
[the remark this answers is not in the document]
Comment (byin001) [reply java/src/main/native/src/cuSpatialJni.cpp:58]: anonymous namespace added

**Review by revans2 — APPROVED** [review]
Looks great

**Review by harrism — CHANGES_REQUESTED** [review]
A few doc edits.

**Review by harrism — APPROVED** [review]

**Pull request merged**